plat/common/include/uk/plat/common/memory.h

```diff
@@ -112,6 +117,80 @@ ukplat_memregion_list_insert(struct ukplat_memregion_list *list,
 	return (int)i;
 }
 
+/**
+ * Insert a new region into memory region list. This extends
+ * ukplat_memregion_list_insert to carve out the area of any pre-existing
+ * overlapping regions. The virtual addresses will leave out any carved out
+ * region.
+ * If there is not enough space for all resulting region, the function will
+ * insert as many as possible and then signal return an error.
+ * @param list
+ *   The memory region list to insert the range into
+ * @param mrd
+ *   The memory range to insert
+ * @param min_size
+ *   The minimum size a inserted region has to have. Regions smaller than this
+ *   size won't be added. Settings this parameter zero will disable this
+ *   behavior.
+ * @returns
+ *   zero if the operation was successful, an negative errno otherwise
+ */
+static inline int
+ukplat_memregion_list_insert_split_phys(struct ukplat_memregion_list *list,
+					const struct ukplat_memregion_desc *mrd,
+					__sz min_size)
+{
+	struct ukplat_memregion_desc *mrdp;
+	struct ukplat_memregion_desc mrdc;
+	__paddr_t pstart, pend;
+	__vaddr_t voffset;
+	int i;
+	int rc;
+
+	voffset = mrd->vbase - mrd->pbase;
+	pstart = mrd->pbase;
+	pend = mrd->pbase + mrd->len;
+
+	memcpy(&mrdc, mrd, sizeof(mrdc));
+
+	for (i = 0; i < (int)list->count; i++) {
+		ukplat_memregion_get(i, &mrdp);
+		if (!ukplat_memregion_desc_overlap(mrdp, pstart, pend))
+			continue;
+
+		if (!mrdp->type)
+			continue;
+
+		if (pstart < mrdp->pbase) {
+			/* Some part of the inserted region is before the
+			 * overlapping region. Try to insert that part if it's
+			 * large enough.
+			 */
+			mrdc.pbase = pstart;
+			mrdc.vbase = pstart + voffset;
+			mrdc.len   = mrdp->pbase - pstart;
+
+			if (min_size && mrdc.len >= min_size) {
+				rc = ukplat_memregion_list_insert(list, &mrdc);
+				if (rc < 0)
+					return -ENOMEM;
```

Comment:
Currently `ukplat_memregion_list_insert` traverses the list (which is actually an array) from the tail to the head to find the right index at which the insertion should happen.

The thing is that we already know the position at which we want to insert the element: `i`. 
Not sure if we really want to do it in this PR, as it's already getting large. Still, adding a TODO as a comment would be nice.

Reply:
On a different note, currently we are traversing 2 times the same element whenever an insertion happens:

Let's say i = 0 and we want to insert a new element. The element will be inserted as the first element, and the current `mrdp` will become the second in the list. In the next iteration, we will traverse the same `mrdp`.

I think here we should increment i by 1.